mopidy_spotify/playlists.py

```diff
@@ -1,90 +1,74 @@
     def __init__(self, backend):
         self._backend = backend
         self._timeout = self._backend._config["spotify"]["timeout"]
+        self._loaded = False
 
     def as_list(self):
-        with utils.time_logger("playlists.as_list()"):
+        with utils.time_logger("playlists.as_list()", logging.INFO):
+            if not self._loaded:
+                return []
+
             return list(self._get_flattened_playlist_refs())
 
     def _get_flattened_playlist_refs(self):
-        if self._backend._session is None:
-            return
-
-        if self._backend._session.playlist_container is None:
+        if not self._backend._web_client.logged_in:
             return
 
-        username = self._backend._session.user_name
-        folders = []
-
-        for sp_playlist in self._backend._session.playlist_container:
-            if isinstance(sp_playlist, spotify.PlaylistFolder):
-                if sp_playlist.type is spotify.PlaylistType.START_FOLDER:
-                    folders.append(sp_playlist.name)
-                elif sp_playlist.type is spotify.PlaylistType.END_FOLDER:
-                    folders.pop()
-                continue
-
+        web_client = self._backend._web_client
+        for web_playlist in web_client.get_user_playlists():
             playlist_ref = translator.to_playlist_ref(
-                sp_playlist, folders=folders, username=username
+                web_playlist, web_client.user_id
             )
             if playlist_ref is not None:
                 yield playlist_ref
 
     def get_items(self, uri):
-        with utils.time_logger(f"playlist.get_items({uri})"):
+        with utils.time_logger(f"playlist.get_items({uri})", logging.INFO):
             return self._get_playlist(uri, as_items=True)
 
     def lookup(self, uri):
-        with utils.time_logger(f"playlists.lookup({uri})"):
+        with utils.time_logger(f"playlists.lookup({uri})", logging.DEBUG):
             return self._get_playlist(uri)
 
     def _get_playlist(self, uri, as_items=False):
-        try:
-            sp_playlist = self._backend._session.get_playlist(uri)
-        except spotify.Error as exc:
-            logger.debug(f"Failed to lookup Spotify URI {uri}: {exc}")
-            return
-
-        if not sp_playlist.is_loaded:
-            logger.debug(f"Waiting for Spotify playlist to load: {sp_playlist}")
-            sp_playlist.load(self._timeout)
-
-        username = self._backend._session.user_name
-        return translator.to_playlist(
-            sp_playlist,
-            username=username,
-            bitrate=self._backend._bitrate,
-            as_items=as_items,
+        return playlist_lookup(
+            self._backend._session,
+            self._backend._web_client,
+            uri,
+            self._backend._bitrate,
+            as_items,
         )
 
     def refresh(self):
-        pass  # Not needed as long as we don't cache anything.
+        if not self._backend._web_client.logged_in:
```

Comment:
Should I also set `self._loaded = False` above this?

Reply:
If we were able to process new requests while we refresh, it would make sense to continue serving the playlists we already have for concurrent requests while refreshing. Setting `self._loaded = False` would make those imaginary requests get an empty response instead. Or have I missed something?

Reply:
I was thinking we'd then potentially be modifying the cached data (when it's dirty) in multiple places concurrently. Which night lead to some unexpected results.

Reply:
Yeah, we're in an actor here, so we have by definition only one request to process at a time. Thus, if you set `self._loaded = False` at the beginning of the method should have no effect, as no other code will run and be able to read it before you set it to `True` at the end of the method.